src/Raven.Server/Documents/Subscriptions/AbstractSubscriptionStorage.cs

```diff
@@ -0,0 +1,171 @@
+    protected readonly ServerStore _serverStore;
+    protected string _databaseName;
+    private readonly SemaphoreSlim _concurrentConnectionsSemiSemaphore;
+    protected Logger _logger;
+
+    protected AbstractSubscriptionStorage(ServerStore serverStore, int maxNumberOfConcurrentConnections)
+    {
+        _serverStore = serverStore;
+        _concurrentConnectionsSemiSemaphore = new SemaphoreSlim(maxNumberOfConcurrentConnections);
+        LowMemoryNotification.Instance.RegisterLowMemoryHandler(this);
+    }
+
+    public abstract void Initialize(string name);
+    protected abstract void DropSubscriptionConnections(TState state, SubscriptionException ex);
+    protected abstract void SetConnectionException(TState state, SubscriptionException ex);
+    protected abstract string GetSubscriptionResponsibleNode(DatabaseRecord databaseRecord, SubscriptionState taskStatus);
+    protected abstract bool SubscriptionChangeVectorHasChanges(TState state, SubscriptionState taskStatus);
+
+    public bool DropSubscriptionConnections(long subscriptionId, SubscriptionException ex)
+    {
+        // TODO: egor switch it back to TryGetValue (like in v5.4), or need to dispose it
+        if (_subscriptions.TryRemove(subscriptionId, out TState state) == false)
+            return false;
+
+        DropSubscriptionConnections(state, ex);
+
+        if (_logger.IsInfoEnabled)
+            _logger.Info($"Subscription with id '{subscriptionId}' and name '{state.SubscriptionName}' connections were dropped.", ex);
+
+        return true;
+    }
+
+    
+    private bool DeleteAndSetException(long subscriptionId, SubscriptionException ex)
+    {
+        if (_subscriptions.TryRemove(subscriptionId, out TState state) == false)
+            return false;
+
+        SetConnectionException(state, ex);
+
+        state.Dispose();
+
+        if (_logger.IsInfoEnabled)
+            _logger.Info($"Subscription with id '{subscriptionId}' and name '{state.SubscriptionName}' was deleted and connections were dropped.", ex);
+
+        return true;
+    }
+
+    public virtual void HandleDatabaseRecordChange(DatabaseRecord databaseRecord)
+    {
+        using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+        using (context.OpenReadTransaction())
+        {
+            //checks which subscriptions should be dropped because of the database record change
+            foreach (var subscriptionStateKvp in _subscriptions)
+            {
+                var subscriptionName = subscriptionStateKvp.Value.SubscriptionName;
+                if (subscriptionName == null)
+                    continue;
+
+                var id = subscriptionStateKvp.Key;
+                var subscriptionConnectionsState = subscriptionStateKvp.Value;
+
+                using var subscriptionStateRaw = _serverStore.Cluster.Subscriptions.ReadSubscriptionStateRaw(context, _databaseName, subscriptionName);
```

Comment:
the `_databaseName` here for sharded might be `Northwind$0` right? is that okay here?

Reply:
IIRC, name comes from Initalize method in which we are passing the appropriate name (without $) - @garayx please confirm. Also you can probably now move this 'name' to ctor as I wrote in other comment.

Reply:
> the `_databaseName` here for sharded might be `Northwind$0` right? is that okay here?

no `this is full name for sharded db `